File: frontend/src/lib/overrange.ts
/**
 * Detects over-range pixels in TemperatureLinear Mono16 counts: pixels the camera could not
 * represent because the scene exceeded its range.
 *   - saturated: count at/above the 16-bit ceiling (scene hotter than the camera's range);
 *   - wrapped: a count that overflowed 16 bits and now reads far colder than physically possible
 *     (the "hot region shows as cold" artefact). Wrapped pixels are only flagged when the frame is
 *     clearly hot (its peak is well into the upper range), so a genuinely cold scene is never
 *     touched. The true temperature of these pixels is unrecoverable — flagging keeps them from
 *     rendering as false cold and from poisoning ROI statistics and the auto colour range.
 */
export const SAT_HI = 65000; // near uint16 max
const FLOOR = 26000; // ~ -13 °C in 10 mK counts: implausibly cold in a hot scene → wrapped
const HOT_MAX = 45000; // ~ 177 °C: the frame is hot enough for over-range/wrap to be plausible

export interface OverRange { mask: Uint8Array; saturated: number; wrapped: number; }

export function overRangeMask(counts: Uint16Array, _w: number, _h: number): OverRange | null {
  let saturated = 0, wrapped = 0, maxCount = 0;
  const mask = new Uint8Array(counts.length);
  for (let i = 0; i < counts.length; i++) {
    const v = counts[i];
    if (v > maxCount) maxCount = v;
    if (v >= SAT_HI) { mask[i] = 1; saturated++; }
  }
  if (maxCount >= HOT_MAX) {
    for (let i = 0; i < counts.length; i++) {
      if (mask[i]) continue;
      if (counts[i] <= FLOOR) { mask[i] = 1; wrapped++; }
    }
  }
  return saturated || wrapped ? { mask, saturated, wrapped } : null;
}

File: backend/flir_research_interface/radiometry/overrange.py
"""Over-range (saturated / 16-bit-wrapped) pixel detection for TemperatureLinear counts.

When a scene exceeds the camera's range the hottest counts saturate at the 16-bit ceiling or wrap
around and read as extreme cold. Such pixels carry no valid temperature; excluding them keeps
whole-frame and ROI statistics (and the auto color range) from being poisoned by false values.
Mirrors ``frontend/src/lib/overrange.ts``.
"""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

SAT_HI = 65000  # near uint16 max: saturated
_FLOOR = 26000  # ~ -13 C in 10 mK counts: implausibly cold in a hot scene -> wrapped
_HOT_MAX = 45000  # ~ 177 C: the frame is hot enough for over-range to be plausible


def over_range_mask(counts: npt.NDArray[np.uint16]) -> npt.NDArray[np.bool_] | None:
    """Boolean mask of over-range pixels, or ``None`` when the frame has none."""
    mask = counts >= SAT_HI
    if int(counts.max(initial=0)) >= _HOT_MAX:
        mask = mask | (counts <= _FLOOR)
    return mask if bool(mask.any()) else None
